src/pyhf/infer/calculators.py

```diff
@@ -0,0 +1,165 @@
+    In the ROOT implementation the same sigma is assumed for both hypotheses
+    and :math:`p`-values etc are computed in that space.
+    This assumption is necessarily valid, but we keep this for compatibility reasons.
+
+    In the :math:`-\hat{\mu}/\sigma` space, the test statistic (i.e. :math:`\hat{\mu}/\sigma`) is
+    normally distributed with unit variance and its mean at
+    the :math:`-\mu'`, where :math:`\mu'` is the true poi value of the hypothesis.
+    """
+
+    def __init__(self, shift):
+        """
+        Asymptotic test statistic distribution.
+
+        Args:
+            shift (`float`): The displacement of the test statistic distribution.
+
+        Returns:
+            ~pyhf.infer.calculators.AsymptoticTestStatDistribution: The asymptotic distribution of test statistic.
+
+        """
+        self.shift = shift
+
+    def pvalue(self, value):
+        """
+        Compute the :math:`p`-value for a given value of the test statistic.
+
+        Args:
+            value (`float`): The test statistic value.
+
+        Returns:
+            Float: The integrated probability to observe a value at least as large as the observed one.
+
+        """
+        tensorlib, _ = get_backend()
+        return 1 - tensorlib.normal_cdf(value - self.shift)
+
+    def expected_value(self, nsigma):
+        """
+        Return the expected value of the test statistic.
+
+        Args:
+            nsigma (`int` or `tensor`): The number of standard deviations.
+
+        Returns:
+            Float: The expected value of the test statistic.
+        """
+        return nsigma
+
+
+class AsymptoticCalculator(object):
+    """The Asymptotic Calculator."""
+
+    def __init__(self, data, pdf, init_pars=None, par_bounds=None, qtilde=False):
+        """
+        Asymptotic Calculator.
+
+        Args:
+            data (`tensor`): The observed data.
+            pdf (~pyhf.pdf.Model): The statistical model adhering to the schema ``model.json``.
+            init_pars (`tensor`): The initial parameter values to be used for fitting.
+            par_bounds (`tensor`): The parameter value bounds to be used for fitting.
+
+        Returns:
+            ~pyhf.infer.calculators.AsymptoticCalculator: The calculator for asymptotic quantities.
+
+        """
+        self.data = data
+        self.pdf = pdf
+        self.init_pars = init_pars or pdf.config.suggested_init()
+        self.par_bounds = par_bounds or pdf.config.suggested_bounds()
+        self.qtilde = qtilde
+
+    def distributions(self, poi_test):
+        """
+        Probability Distributions of the test statistic value under the signal + background and and background-only hypothesis.
+
+        Args:
+            poi_test: The value for the parameter of interest.
+
+        Returns:
+            Tuple (~pyhf.infer.calculators.AsymptoticTestStatDistribution): The distributions under the hypotheses.
+
+        """
+        sb_dist = AsymptoticTestStatDistribution(-self.sqrtqmuA_v)
```

Comment:
this only works if someone calls `teststatistic` before this function because `self.sqrtqmuA_v` won't be defined.

Reply:
yes